ext/ffi/fast_call.rs

```diff
@@ -0,0 +1,188 @@
+}
+
+pub(crate) fn make_template(sym: &Symbol, trampoline: &Trampoline) -> Template {
+  let args = once(fast_api::Type::V8Value)
+    .chain(sym.parameter_types.iter().map(|t| t.into()))
+    .collect::<Vec<_>>();
+
+  Template {
+    args: args.into_boxed_slice(),
+    ret: (&fast_api::Type::from(&sym.result_type)).into(),
+    symbol_ptr: trampoline.ptr(),
+  }
+}
+
+/// Trampoline for fast-call FFI functions
+///
+/// Removes first argument (Javascript object receiver) and shifts the rest of arguments to the left
+pub(crate) struct Trampoline(ExecutableBuffer);
+
+impl Trampoline {
+  fn ptr(&self) -> *const c_void {
+    &self.0[0] as *const u8 as *const c_void
+  }
+
+  fn compile(sym: &Symbol) -> Self {
+    #[derive(Clone, Copy)]
+    struct ArgCounter {
+      integer: i32,
+      sse: i32,
+    }
+    let mut counter = ArgCounter {
+      integer: -6,
+      sse: -8,
+    };
+    let args_to_shift: Vec<_> = sym
+      .parameter_types
+      .iter()
+      .filter_map(|p| match p {
+        NativeType::F32 | NativeType::F64 => {
+          counter.sse += 1;
+          if counter.sse > 0 && counter.integer >= 0 {
+            // floats are only shifted to accomodate integer shift in the stack
+            Some(counter)
+          } else {
+            None
+          }
+        }
+        _ => {
+          counter.integer += 1;
+          Some(counter)
+        }
+      })
+      .collect();
+
+    let mut ops = dynasmrt::x64::Assembler::new().unwrap();
+    if args_to_shift.is_empty() {
+      dynasm!(ops
+        ; .arch x64
+        ; xor rdi, rdi
+      );
+    } else {
+      for arg in args_to_shift {
+        match arg {
+          ArgCounter { integer: -5, .. } => dynasm!(ops
+            ; .arch x64
+            ; mov rdi, rsi
+          ),
+          ArgCounter { integer: -4, .. } => dynasm!(ops
+            ; .arch x64
+            ; mov rsi, rdx
+          ),
+          ArgCounter { integer: -3, .. } => dynasm!(ops
+            ; .arch x64
+            ; mov rdx, rcx
+          ),
+          ArgCounter { integer: -2, .. } => dynasm!(ops
+            ; .arch x64
+            ; mov rcx, r8
+          ),
+          ArgCounter { integer: -1, .. } => dynasm!(ops
+            ; .arch x64
+            ; mov r8, r9
+          ),
+          ArgCounter { integer: 0, sse } => dynasm!(ops
+            ; .arch x64
+            ; mov r9, [rsp + (1 + sse.max(0)) * 8]
+          ),
+          ArgCounter { integer, sse } => dynasm!(ops
+            ; .arch x64
+            ; mov rax, [rsp + (1 + integer + sse.max(0)) * 8]
+            ; mov [rsp + (integer + sse.max(0)) * 8], rax
+          ),
+        }
+      }
+    }
+
+    // tail-call
+    // SAFETY: stack pointer is not modified, therefore stack remains (un)aligned.
+    // Stack contains one extra parameter (the last one) when using >6 parameters, which the function will ignore
+    dynasm!(ops
+      ; .arch x64
+      ; mov rax, QWORD sym.ptr.as_ptr() as _
```

Comment:
Same thing here: For 8 / 16 bit integer return values, are we sure that it's okay to just `mov` them over without some explicit conversion to 32 bit integers?

Reply:
same as https://github.com/denoland/deno/pull/15305#discussion_r929391251

Reply:
@Phosra had a [good comment](https://github.com/denoland/deno/pull/15122#discussion_r917323197) related to this stuff. See also the linked [Rust bug](https://github.com/rust-lang/rust/issues/97463) and further a [PR to fix it](https://github.com/rust-lang/rust/pull/97800).

This is not something that can just be handwaved away. At present, the code is essentially creating undefined behaviour by passing 32 bit integers as 8 or 16 bit integers without zero or sign extending them to the 32 bit boundary that, as you said, is kind of a de facto standard. It is also not taking into account possible garbage bits in 8 or 16 bit integer return values, which there may well be as there is no de facto standard on if a return value should be zero or sign extended before returning. If anything, the standard is that no such actions are taken, garbage bits are left as they were.

Reply:
settled in #15305 .

Thanks for the links!